presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergAbstractMetadata.java

```diff
@@ -432,6 +438,98 @@ public Optional<SystemTable> getSystemTable(ConnectorSession session, SchemaTabl
+    {
+        if (!tableLayoutHandle.isPresent()) {
+            return true;
+        }
+
+        // Allow metadata delete for range filters on partition columns.
+        IcebergTableLayoutHandle layoutHandle = (IcebergTableLayoutHandle) tableLayoutHandle.get();
+
+        TupleDomain<ColumnHandle> domainPredicate = layoutHandle.getTupleDomain();
+        if (domainPredicate.isAll()) {
+            return true;
+        }
+
+        Set<String> predicateColumnNames = domainPredicate.getDomains().get().keySet().stream()
+                .map(columnHandle -> ((IcebergColumnHandle) columnHandle).getName())
+                .collect(toImmutableSet());
+
+        IcebergTableHandle handle = (IcebergTableHandle) tableHandle;
+        Table icebergTable = getIcebergTable(session, handle.getSchemaTableName());
+
+        boolean canSupportsMetadataDelete = true;
+        for (PartitionSpec spec : icebergTable.specs().values()) {
+            // Currently we do not support delete when any partition columns in predicate is not transform by identity()
+            Set<String> partitionColumnNames = spec.fields().stream()
+                    .filter(field -> field.transform().isIdentity())
+                    .map(field -> field.name()).collect(Collectors.toSet());
+
+            if (!partitionColumnNames.containsAll(predicateColumnNames)) {
+                canSupportsMetadataDelete = false;
+                break;
+            }
+        }
+
+        return canSupportsMetadataDelete;
+    }
+
+    @Override
+    public OptionalLong metadataDelete(ConnectorSession session, ConnectorTableHandle tableHandle, ConnectorTableLayoutHandle tableLayoutHandle)
+    {
+        IcebergTableHandle handle = (IcebergTableHandle) tableHandle;
+        IcebergTableLayoutHandle layoutHandle = (IcebergTableLayoutHandle) tableLayoutHandle;
+
+        Table icebergTable;
+        try {
+            icebergTable = getIcebergTable(session, handle.getSchemaTableName());
+        }
+        catch (Exception e) {
+            throw new TableNotFoundException(handle.getSchemaTableName());
+        }
+
+        TupleDomain<ColumnHandle> domainPredicate = layoutHandle.getTupleDomain();
+        if (domainPredicate.isAll()) {
+            try (CloseableIterable<FileScanTask> files = icebergTable.newScan().planFiles()) {
+                return OptionalLong.of(removeScanFiles(icebergTable, files));
+            }
+            catch (IOException e) {
+                throw new PrestoException(GENERIC_INTERNAL_ERROR, "failed to scan files for delete", e);
+            }
+        }
+        else {
+            Expression filterExpression = toIcebergExpression(handle.getPredicate());
+            try (CloseableIterable<FileScanTask> files = icebergTable.newScan()
+                    .filter(filterExpression).useSnapshot(handle.getSnapshotId().get()).planFiles()) {
```

Comment:
So it seems like you've considered the case where a user tries to delete metadata from the snapshot of a table, but I don't fully understand the semantics of how that would work. If you delete data in older snapshots that doesn't propagate to newer snapshots does it?

e.g. you insert record 1 and record 2 in snapshot A, you insert record 3 in snapshot B. All records are in the same partition. Then you run `DELETE FROM "<table>@a" WHERE ....` for  snapshot C. This deletes the partition containing records 1, 2, and 3. However, In snapshot C does this mean when you query you will only get record 3, or do you get no records?

I have a few questions I'd like to answer:

1. If we do support deleting partitions from a particular snapshot, what are the correct semantics? Should we document it? Does iceberg document it?
2. Otherwise, if there isn't a clear definition from Iceberg, personally I think we could disallow metadata deletes in the case a user queries an iceberg table like `"orders@<snapshot>"`. What is your opinion?
3. In a similar vein, how about partition evolution, and that in combination with querying at specific snapshots? Will this handle the case where you try to delete a partition which was not yet created/specified at a particular snapshot?

If we do support all of these case (deleting at particular snapshots + partition evolution), then I think we should also add some tests for them.

Reply:
Great questions. Originally I did not consider metadata deletes with user given snapshot, just sorry for overlooked this kind of usage. But now I think we could support deleting partitions from a particular snapshot, considering iceberg's specification about metadata and snapshot. In your example, record 3 should be the remaining record, because even in the same partition, data of different snapshots would be maintained in different files. I will do some tests to confirm that.

As for partition evolution, in current implementation, the requirements for supporting metadata deletes has been checked for each partition specs, I think that could make sure there is no problem. I will add some tests for it.

Reply:
Sounds good! Please let me know if you find anything in the Iceberg docs

Reply:
Iceberg specification do not define such behavior. And in iceberg-api, it's interface DeleteFiles declares that all it's behavior of file deletions would be applied to the latest table snapshot. So, I think it's better to explicitly disallow user given snapshot for metadata deletes. Such behavior may lead in confusion, for example, when we try to delete some rows in an older snapshot which are modified or deleted in newer snapshots.

As for partition evolution, it seems like we haven't consider partition spec's evolution when try to alter table,
for example rename column, add column or drop column. So I plan to make some PRs to support that. When it's done, we can add test cases for partition evolution.

Others are all fixed, please take a look.